### DISCUSSION TITLE
Linear Actuator, Contact-based Control and Rigid Constraint of the End-effector

### DISCUSSION THREAD

**Discussion by ozilxu** [SofaDefrost/SoftRobots#297] [Help - Programming with SoftRobots] [answered]
Hi SOFA developer,  @alxbilger @ErwanDouaille 

Recently, I wanted to build a parallel trunk robot. May I ask several questions about the development? I have tried to search for the solution in the example code, but I didn't find something related (or maybe I don't have a full understanding of the SOFA code).

1. How do we implement a linear actuator for the trunk robot? Since the base of this robot is fixed, how can I add the "linear actuation" to the base to allow the robot to have linear displacement (both forward and inverse mode)?

2. How do we add the contact of the environment so that the trunk minimizes the contact force between the robot and the environment?

3. If I add an end-effector constraint to the end of the trunk to make it like a delta parallel robot. How can I define such a rigid constraint (fixed constraints and ball-joints constraints)?

Many thanks for your answers in advance!

**Comment by peyronq1** [comment]
Hello @ozilxu, :)

1. For linear actuator, if you want to do direct simulations, you can fix the position of the base of your leg at the desired one, with a FixedProjectiveConstraints as in the example of parallel manipulator here (https://github.com/SofaDefrost/TIRREX.Migribot). For inverse simulation you can look at the SoftRobots.Inverse examples (https://github.com/SofaDefrost/SoftRobots.Inverse/tree/main/examples/sofapython3/component/constraint). There is a SlidingActuator that can be interesting for you I think.

2. I am not sure there is already a feature to minimize the total contact force along with solving the inverse model... You might have to implement your own Sofa controller to do so.

3. Can you clarify your question, please? Are you speaking about the constraint of assembling the different legs on the platform?

**Reply by ozilxu** [reply]
Hi Quentin, 

Thanks a lot for your detailed answers! I will check the scene of the Migribot and the sliding actuators!

For contact-based control, I think I need to define the collision in the scene and let the QP solver compute the actuation input.

For the last question, yes, constraining different "straight" legs is easy. But if the leg is straight at the beginning, they will be "pre-curved" after adding a different end-effector. I am unsure whether it is feasible to constrain or assemble different pre-curved legs on the platform. Many thanks!

**Reply by peyronq1** [reply, marked as answer]
For the contact-based control, what you propose won't be enough from what I understood. The QP inverse solver minimizes the distance between the actual and desired effector positions while satisfying the static/dynamic model of the robot as a constraint. The cost function does not include the mean/max/total contact force applied to the environment. As a consequence, when there is a contact, the QP will find actuator values such that the energy in the robot is minimal, but won't minimize the contact force. 

For the assembly, if I understand well, your problem is how to pass from the initial configuration of the legs, without the platform constraint, to the initial assembled configuration where the legs are deformed, is that right? Do you have a problem with this because it does not converge when you define your legs straight and you just hit the run button?

**Reply by ozilxu** [reply]
Thanks for the detailed answer!

For the contact-based control, yes you are right, it might need to define the constraint between the robot and the environment such as complimentary constraints following this paper: [Optimization-based inverse model of soft robots with contact handling](https://ieeexplore.ieee.org/abstract/document/7856996). The optimization problem must be solved with the constraint.  I am not so sure whether this complimentary constraint is already "built" or "developed" in the QP solver from the SoftRobot Plugins. 

For the leg pre-curvature, I think the other scene might suit the legs' "pre-curvature": https://github.com/SofaDefrost/TIRREX.MicroParallel. The leg is straight at the beginning, and it has an "initial assembled configuration" after constraining the end of the legs. I tried to run this scene, and it fits the curvature problem so well.
